llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

```diff
@@ -4000,12 +4000,14 @@ static bool isReverseOrder(ArrayRef<unsigned> Order) {
 
 /// Checks if the given array of loads can be represented as a vectorized,
 /// scatter or just simple gather.
-static LoadsState canVectorizeLoads(ArrayRef<Value *> VL, const Value *VL0,
+static LoadsState canVectorizeLoads(const BoUpSLP &R, ArrayRef<Value *> VL,
```

Comment:
my 2 cents:  looks like its would be much simpler to make it a BoUpSLP method rather than pump many of of these args thru a call.  Whole bunch of resources for analysis  would be readily available if it was a class method.